rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+- Feature Name: `abi_errors`
+- Start Date: 2024-11-18
+- RFC PR: [FuelLabs/sway-rfcs#43](https://github.com/FuelLabs/sway-rfcs/pull/43)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to standardize error handling in Sway so that there is a
+deliberate and clear way to produce error messages that are easily consumable by
+language tools and do not burden on chain artifacts with error strings.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, reverting a sway program only produces an opaque error code.
+Sometimes this code is unusual enough that with the source code of the contract
+it is possible to figure out what is going on, but a lot of the time it's a
+meaningless number that only means that some failure happened somewhere.
+
+It's not reasonable to expect people to be able to debug sway programs in these
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings minimally burden the final bytecode size.
```

Comment:
I'm having trouble understanding this sentence,  does the suggestion below clarify the intention?

```suggestion
defined types and error strings to minimize the final bytecode size.
```

Reply:
Hmmm, I find the "minimally burden" better explaining the intention :smile: We want the error strings to be available at the end (somehow, via tooling), but not to burden the bytecode with them.